packages/core/src/model.ts

```diff
@@ -80,23 +110,122 @@ export class ChartModel {
 	 */
 	setData(newData) {
 		const sanitizedData = this.sanitize(Tools.clone(newData));
-		const dataLabels = this.generateDataLabels(sanitizedData);
+		const dataGroups = this.generateDataGroups(sanitizedData);
 
 		this.set({
 			data: sanitizedData,
-			dataLabels
+			dataGroups
 		});
 
 		return sanitizedData;
 	}
 
-	generateDataLabels(newData) {
-		const dataLabels = {};
-		newData.datasets.forEach(dataset => {
-			dataLabels[dataset.label] = Configuration.legend.items.status.ACTIVE;
+	/*
+	 * Data groups
+	*/
+	protected updateAllDataGroups() {
+		// If allDataGroups hasn't been initialized yet
+		// Set it to the current set of data groups
+		if (!this.allDataGroups) {
+			this.allDataGroups = this.getDataGroups().map(group => group.name);
+		} else {
+			// Loop through current data groups
+			this.getDataGroups().forEach(dataGroup => {
+				// If group name hasn't been stored yet, store it
+				if (this.allDataGroups.indexOf(dataGroup.name) === -1) {
+					this.allDataGroups.push(dataGroup.name);
+				}
+			});
+		}
+	}
+
+	protected generateDataGroups(data) {
+		const { groupIdentifier } = this.getOptions().data;
+		const { ACTIVE } = Configuration.legend.items.status;
+
+		const uniqueDataGroups = map(data, datum => datum[groupIdentifier]).keys();
+		return uniqueDataGroups.map(groupName => ({
+			name: groupName,
+			status: ACTIVE
+		}));
+	}
+
+	getDataGroups() {
+		return this.get("dataGroups");
+	}
+
+
+	getDataGroupNames() {
+		return this.get("dataGroups").map(dataGroup => dataGroup.name);
```

Comment:
```suggestion
		return this.getDataGroups().map(dataGroup => dataGroup.name);
```
might make sense, conceptually